packages/@glimmer/runtime/lib/utils/tags.ts

```diff
@@ -12,7 +11,7 @@ import { Tag, Tagged, createCombinatorTag, CONSTANT_TAG } from '@glimmer/validat
 export function combineTagged(tagged: ReadonlyArray<Tagged>): Tag {
   let optimized: Tag[] = [];
 
-  for (let i = 0, l = tagged.length; i < l; i++) {
```

Comment:
I'm curious about this one, why remove the length caching?

Reply:
@krisselden showed me [this very helpful blog post](https://mrale.ph/blog/2014/12/24/array-length-caching.html) that explained how caching actually ends up forcing the VM to do _more_ work, not less. 

Basically, array access needs to be bounds checked no matter what, so `.length` is always read anyways. If it is read as part of the `for` loop declaration, then the VM is smart about it and caches the value for when the array is accessed in the loop. Otherwise, it has to load it again itself when doing the check.